keps/draft-20180426-security-conformance.md

```diff
@@ -0,0 +1,682 @@
+
+### Security Conformance Profiles
+
+The criteria are categorized into a few security conformance profiles,
+based on the following factors:
+
+- The source of threats:
+  - Use of Kubernetes APIs:
+    The attack tries to compromise the cluster using Kubernetes APIs,
+    compromised user identity is out-of-scope.
+  - Host access on Kubernetes masters and nodes:
+    The attack accesses the hosts running master/kubelet directly.
+    Access with privileged user, like root, or sudo users,
+    or the user/group running Kubernetes components, is out-of-scope of this proposal.
+  - Access prohibited information (e.g. secrets):
+    The attack tries to access prohibited information via Kubernetes APIs,
+    or direct access to the host file system,
+    or direct access to etcd.
+  - Malicious workloads (not through Kubernetes APIs):
+    The workload gains privileges/capabilities, like SYSADMIN capability,
+    use of host namespaces, access to raw sockets, etc., to compromise the system.
+
+- The components to be protected from threats:
+  - The Kubernetes cluster (not including workloads)
+  - The workloads grouped by certain scope (like tenant, namespace, pods)
+
+
+The profiles:
+
+- `Essential` (aka `Trusted`): the baseline security requirements for the cluster.
+
+  It covers the threat sources of:
+  - Use of Kubernetes APIs
+  - Host access on Kubernetes masters and nodes
+  - Access prohibited information
+
+  It protects the components of:
+  - Kubernetes cluster
+
+  This profile focuses on minimum security requirements,
+  and provides maximum compatibility/capabilities to workloads.
+  It assumes the workloads and authorized users are trusted.
+  It's best for the use case to operate a secure Kubernetes cluster and
+  share internally with trusted users (e.g. an internal engineering team).
+
+  The criteria in this category must be satisfied for a Kubernetes cluster
+  to be considered secure at minimal.
+
+- `Restricted` (aka `Untrusted`):
+
+  In addition to `Essential`, it covers the threat sources of:
+  - Malicious workloads (not through Kubernetes API)
+
+  This profile limits the compatibility/capabilities to workloads to prevent
+  them from compromising the cluster.
+  It's still possible a malicious workload will compromise other workloads.
+  It's best for the use case to operate a secure Kubernetes cluster and share
+  with untrusted users without taking multi-tenancy into consideration.
+
+- `Isolated` (aka `Multi-tenancy`):
+
+  In addition to `Restricted`, it protects workloads from being compromised by
+  other workloads across the boundary of another tenant.
+
+  It's for the use case supporting multi-tenancy by sharing a Kubernetes cluster.
+
+- `Sandboxed`:
+
+  Similar to `Isolated`, but using a different container runtime for better isolation,
+  and possibly offering better compatibility and capabilities.
+  For example, using hypervisors, gVisor etc. which provides an additional layer of protection.
```

Comment:
How are you going to test this? We've been talking about how to define sandboxes concretely, and that's proving to be difficult on its own. I'm not sure it's going to be possible to test for those properties.